grudge/dt_utils.py

```diff
@@ -45,22 +57,59 @@
 from pytools import memoize_on_first_arg
 
 
+def characteristic_lengthscales(dcoll: DiscretizationCollection) -> DOFArray:
+    r"""Computes the characteristic lengthscales at each node of a physical element.
+    The estimate is obtained using the following formula:
+
+    .. math::
+
+        h_{loc} = \operatorname{min}\left(\Delta r_i\right) r_D
+
+    where :math:`\operatorname{min}\left(\Delta r_i\right)` is the minimum
+    node distance on the reference cell (see :func:`dt_non_geometric_factors`),
+    and :math:`r_D` is the inradius of the cell (see :func:`dt_geometric_factors`).
+
+    :returns: a :class:`~meshmode.dof_array.DOFArray` containing a characteristic
+        lengthscale for each element, at each nodal location.
+
+    .. note::
+
+        While a prediction of stability is only meaningful in relation to a given
+        time integrator with a known stability region, the lengthscale provided here
+        is not intended to be specific to any one time integrator, though the
+        stability region of standard four-stage, fourth-order Runge-Kutta
+        methods has been used as a guide. Any concrete time integrator will
+        likely require scaling of the values returned by this routine.
+    """
+    return DOFArray(
+        dcoll._setup_actx,
+        data=tuple(
+            # Scale each group array of geometric factors by the
+            # corresponding group non-geometric factor
+            cng * geo_facts
+            for cng, geo_facts in zip(dt_non_geometric_factors(dcoll),
+                                      dt_geometric_factors(dcoll))
```

Comment:
`dt_geometric_factors` also returns unfrozen data in the same way. I propose that we fix this in this PR.

x-ref: https://github.com/inducer/arraycontext/issues/19